File: scripts/_probe/derive_sub_imm.js
// Body-extend-002 derivation: independent compute of expected bytes
// for `0x61 0x50 0x03` (SUB S[0x50] -= 3) at H_2F.
// JS path: loadState(0x50,rax) + subImmRax(3) + storeState(0x50,rax) + 0xC3
const { encodeOp, loadState, storeState } = require('../../yoyo-js/src/platform/encode-x64');
const ls = loadState(0x50, 0, 0);     // slot 0x50*8=640 > 127 → disp32
const ss = storeState(0x50, 0, 0);     // disp32
// 0x61 = SUB slot,imm path. SUB imm8 = 48 83 e8 03 (REX.W + 83 /5 + imm8)
const out = [...ls, 0x48, 0x83, 0xE8, 0x03, ...ss, 0xC3];
console.log('JS stream:', Buffer.from(out).toString('hex'));
console.log('JS len:', out.length);
console.log('JS components:');
console.log('  load  S[0x50]:', Buffer.from(ls).toString('hex'));
console.log('  sub imm=3   :', '48 83 e8 03');
console.log('  store S[0x50]:', Buffer.from(ss).toString('hex'));
console.log('  ret         : c3');

File: scripts/_probe/derive_sub_imm.rs
// Body-extend-002 Rust independent derivation.
// Use the public assembler API to compute bytes for `0x61 0x50 0x03`.
// = emit_sub_imm(0x50, 3) + ret()
// = load_state(0x50, rax) + sub_imm(rax, 3) + store_state(0x50, rax) + ret()
use yoyo_verifier::assembler::{emit_sub_imm, load_state, store_state, sub_imm, ret, Reg};

fn main() {
    let mut v = Vec::new();
    v.extend(load_state(0x50, Reg::Rax).expect("load_state"));
    v.extend(sub_imm(Reg::Rax, 3).expect("sub_imm"));
    v.extend(store_state(0x50, Reg::Rax).expect("store_state"));
    v.extend(ret());
    println!("Rust manual stream: {}", v.iter().map(|b| format!("{:02x}", b)).collect::<String>());
    println!("Rust manual len: {}", v.len());

    let mut v2 = emit_sub_imm(0x50, 3).expect("emit_sub_imm");
    v2.extend(ret());
    println!("Rust helper stream: {}", v2.iter().map(|b| format!("{:02x}", b)).collect::<String>());
    println!("Rust helper len: {}", v2.len());

    assert_eq!(v, v2, "manual vs helper must match");
    println!("Rust: manual ≡ helper ✓");
}
